src/newrelic.ts

```diff
@@ -40,53 +65,361 @@ function addSegment(spec: Spec) {
 
       ctx.actdef.func.$$newrelic_wrapped$$ = true;
     }
-
-    Object.defineProperty(
-      ctx.actdef.func, 'name', { value: 'newrelic_' + origfunc.name })
   }
 }
 
+interface TelemetrySpecMetadata {
+    id: string,
+    tx_id: string,
+    mi_id: string,
+    fullMessage: Nullable<string>,
+    plugin_name?: string,
+    pattern?: string,
+    duration?: string,
+    endTime?: string,
+    res?: string,
+    manualEndTime?: number,
+    manualStartTime?: number,
+    startTime?: string,
+    dispatched?: boolean,
+}
 
-function endSegment(spec: Spec) {
-  const meta = spec.data.meta
-  const context = spec.ctx.seneca.context
-  const endSegmentMap = context.newrelic?.endSegmentMap
-  if (endSegmentMap && endSegmentMap[meta.mi]) {
-    const endSegmentHandler = endSegmentMap[meta.mi].endSegmentHandler
-    if (endSegmentHandler) {
-      delete endSegmentMap[meta.mi]
-      endSegmentHandler()
+interface TelemetryCollector {
+    API_KEY: string,
+    SERVICE_NAME: string,
+    HOST: string,
+    specList: TelemetrySpecMetadata[],
+    extractFullMessage: (spec: any) => Nullable<string>,
+    extractFromSpec: (spec: any, event: any) => TelemetrySpecMetadata,
+    updateSpecList: (specMetadata: TelemetrySpecMetadata) => void,
+    sendTracing: (spec: TelemetrySpecMetadata) => Promise<void | boolean | string>
+}
+
+function TelemetryCollector(apiKey: string, serviceName: string, host: string): TelemetryCollector {
+    return {
+        API_KEY: apiKey,
+        SERVICE_NAME: serviceName,
+        HOST: host,
```

Comment:
We should be using the NewRelic SDK as much as possible for future compatibility - I am pretty sure you don't need to manually send data in this way.
For example: https://github.com/newrelic/newrelic-telemetry-sdk-node/blob/e1b051a8c72ae0cddc1eb30dc6878f65ff8b849a/src/telemetry/spans/client.ts

did you try this: https://github.com/newrelic/newrelic-telemetry-sdk-node#span-overview-and-examples ?